django/db/models/deletion.py

```diff
@@ -51,6 +52,23 @@ def DO_NOTHING(collector, field, sub_objs, using):
     pass
 
 
+def get_candidate_relations_to_delete(opts):
+    # Collect models that contain candidate relations to delete. This may include
+    # relations coming from proxy models.
+    candidate_models = {opts}
+    candidate_models = candidate_models.union(opts.concrete_model._meta.proxied_children)
+    # For each model, get all candidate fields.
+    candidate_model_fields = chain.from_iterable(
+        opts.get_fields(include_hidden=True) for opts in candidate_models
+    )
+    # Keep related objects not coming from ManyToManyFields from all candidate
+    # model fields.
+    return (
+        f for f in candidate_model_fields
+        if f.auto_created and not f.concrete and not f.many_to_many
```

Comment:
On closer inspection - I think @PirosB3's version is pretty close to the simplest correct version here. We're looking for all related_objects that haven't come from a m2m. We can't *just* use `many_to_one and one_to_one` - we need auto_created to make sure we exclude GenericRelation; we also need `not concrete` to exclude proxies.

The only alternate phrasing of this clause that I can see working is:
```
if f.auto_created and not f.concrete and (f.one_to_one or f.many_to_one)
```
which is slightly more explicit about the fields it wants than `not many_to_many`, so I've made that change.